Планирование – залог успеха

AlexBad opened the pull request:
Версия на обсуждение

---
:mortar_board: [Планирование – залог успеха](https://up.htmlacademy.ru/yii/1/user/1540437/tasks/5)

Commit:
Проект базы, и правки Task.php

Commit:
Merge branch 'module1-task3' of github.com:AlexBad/1540437-task-force-1 into module1-task3

AlexBad committed:
Базовая структура базы, изменения в файле task.php

Commit:
Схема, скрин схемы, модель схемы (MySQL Workbench)

Commit:
Merge branch 'module2-task1' of github.com:AlexBad/1540437-task-force-1 into module2-task1

AlexBad committed:
...

Commit:
Вторая версия схемы таблицы

Commit:
Вторая версия базы, на проверку

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`task_requests` (
```
тут больше подойдет **task_responses**, request это все таки - заявка.

AlexBad replied:
Исправил

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+CREATE TABLE IF NOT EXISTS `taskforce`.`categories` (
+  `id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `name` VARCHAR(256) NOT NULL,
+  `rating` INT(11) NOT NULL DEFAULT 0,
```
Что в этом поле хранится?

int64 replied:
И не хватает поля для иконки

AlexBad replied:
> И не хватает поля для иконки

Я всегда оставлял эту логику на приложение. Добавлю в базу, но чувство такое, что оно лишнее в базе.

> Что в этом поле хранится?

Рэйтинг Категории. На сколько популярна эта категория. По задумке, поле должно было +-1 если добавляются\удаляются исполнители с этой категорией, или заказчик просит выполнить задания из этой категории. Немного подумав над вопросом, пришел к выводу, что поле лишнее. Более "правдивые" данные можно получить из связанных таблиц, а если будет трудно для Базы, то можно закинуть в кэш приложения...

int64 replied:
> Я всегда оставлял эту логику на приложение. Добавлю в базу, но чувство такое, что оно лишнее в базе.

Тут логики нет, просто надо где-то хранить путь до изображения.

int64 replied:
> Рэйтинг Категории. 

В ТЗ не нашел такого.

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`user_contacts` (
```
В рамках этого проекта эта таблица излишняя. Лучше добавить поля Телефон, Skype, Telegram в таблицу users.

AlexBad replied:
Убрал таблицу `user_contacts`, перенёс поля в базу users

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `role` ENUM('CUSTOMER', 'PERFORMER') NOT NULL,
+  `name` VARCHAR(245) NOT NULL,
+  `email` VARCHAR(245) NOT NULL,
+  `birt_date` DATE NULL DEFAULT NULL,
```
birt**h**_date

AlexBad replied:
Исправил

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `location_id` INT(10) UNSIGNED NOT NULL,
+  `about` TEXT NULL DEFAULT NULL,
+  `role` ENUM('CUSTOMER', 'PERFORMER') NOT NULL,
+  `name` VARCHAR(245) NOT NULL,
```
Лучше разделить на два поля, first_name и last_name

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `birt_date` DATE NULL DEFAULT NULL,
+  `password_hash` VARCHAR(256) NOT NULL,
+  `token` VARCHAR(256) NULL DEFAULT NULL,
+  `rating` FLOAT(11) NULL DEFAULT NULL,
```
Это вычислимое поле, его не храним в БД.

int64 replied:
Не хватает следующих полей:

- последняя активность пользователя
- для чекбокса «Показывать контакты только заказчику»
- для чекбокса «Не показывать мой профиль»

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`attachments` (
```
Таблица усложнена, по ТЗ тут будут хранится только фотографии работ пользователя.
Также не хватает поля user_id.
И можно добавить поле пути миниатюры.

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`chats` (
```
В имени таблицы неплохо бы отразить отношение к заданиям и более четче обозначить содержимое этой таблицы. В данном случае у нас в этой таблице хранятся сообщения, а не чаты. Такие моменты легко проверять с помощью поля id, достаточно спросить себя идентификатор чего обозначает поле id.
Соответственно, можно таблицу назвать, например, task_messages.

AlexBad replied:
Изначально так и делал... Была таблица Чаты, и была таблица Сообщения в чате.
Потом, убрал... Исправлю на `task_messages`

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `city_region` VARCHAR(256) NULL DEFAULT NULL,
+  `lattitude` FLOAT(11) NULL DEFAULT NULL,
+  `longtitude` FLOAT(11) NULL DEFAULT NULL,
+  `param` JSON NULL DEFAULT NULL COMMENT 'Параметры кординат в JSON формате',
```
Какие параметры тут хранятся?

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`locations` (
```
Таблицу лучше назвать cities. Так как тут будут храниться города для проверки при публикации задания.

AlexBad replied:
Изменил

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `city_name` VARCHAR(256) NULL DEFAULT NULL,
+  `city_region` VARCHAR(256) NULL DEFAULT NULL,
+  `lattitude` FLOAT(11) NULL DEFAULT NULL,
+  `longtitude` FLOAT(11) NULL DEFAULT NULL,
```
Координаты нужно вынести в таблицу заданий.

AlexBad replied:
Вынес

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `category_id` INT(10) UNSIGNED NOT NULL,
+  `user_id` INT(10) UNSIGNED NOT NULL,
+  `user_location_id` INT(10) UNSIGNED NOT NULL,
```
Это поле нужно переименовать в city_id

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `user_location_id` INT(10) UNSIGNED NOT NULL,
+  `title` VARCHAR(256) NULL DEFAULT NULL,
+  `description` TEXT NULL DEFAULT NULL,
+  `addit_info` TEXT NULL DEFAULT NULL,
```
Непонятное поле

AlexBad replied:
Дополнительная информация, 
Возможно имеет смысл переименовать в `task_comment` или без сокращения `additional_info`
Также, можно просто добавить 'Комментарий' к полю в самой таблице.

Как поступить?

int64 replied:
А что там должно хранится? По ТЗ ничего такого не нашел. Или это описание самого задания? Тогда description.

AlexBad replied:
Просто, дополнительная информация для задания, которое может оставить Заказчик для Исполнителя.

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `description` TEXT NULL DEFAULT NULL,
+  `addit_info` TEXT NULL DEFAULT NULL,
+  `address` TEXT NULL DEFAULT NULL,
+  `status` ENUM('NEW', 'CANCELED', 'INPROGRESS', 'COMPLETE', 'FAIL') NULL DEFAULT NULL,
```
Как правило фиксированные статусы хранятся числами. Например, 1 - новое, 2 - в процессе и тд. Для этих целей идеально подходит tinyint. Соответственно, и в классе нужно будет поменять.
Кроме того, задание не может быть ни в каком статусе, в каком-то точно должно.

AlexBad replied:
Как по мне, ENUM более очевидный.
Переделаю на tinyint...

int64 replied:
Ну ок, enum так enum. Это у меня старческое =) когда всё хранили в числах. Enum действительно понятнее.
Но дефолтное значение всё таки задай.

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `budget` DECIMAL(13,4) UNSIGNED NULL DEFAULT NULL,
+  `created_at` TIMESTAMP NOT NULL DEFAULT current_timestamp(),
+  `updated_at` TIMESTAMP NULL DEFAULT NULL ON UPDATE current_timestamp,
+  `start_before` TIMESTAMP NULL DEFAULT NULL,
+  `done_before` TIMESTAMP NULL DEFAULT NULL,
```
Что-то не нашел по ТЗ такого

AlexBad replied:
**Срок исполнения**

Содержимое поля «срок исполнения» должно быть датой в формате ДД.ММ.ГГГГ

int64 replied:
а второе поле для чего?

AlexBad replied:
По логике должно было быть, время начала и время окончания. "Срок исполнения" или "Период в рамках которого нужно выполнить задание"...

По ТЗ, второе поле выглядит лишним.

AlexBad replied:
Заменил поля `start_before`, `done_before` на 1 поле `start_date`

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  PRIMARY KEY (`id`, `category_id`, `user_id`, `user_location_id`),
+  INDEX `fk_task_categories_idx` (`category_id` ASC),
+  INDEX `fk_task_users_idx` (`user_id` ASC, `user_location_id` ASC),
+  CONSTRAINT `fk_tasks_categories1`
```
Зачем единички в названиях ограничений?

AlexBad replied:
Я пользуюсь MySQL WorkBench... Эта зараза запомнила что уже была таблица с этим названием, и считаем что это вторая таблица с темже именем и добавляет единицу...
В общем, я смотрю чтобы не было единичек, но иногда не замечаю... Постараюсь быть более внимательным.

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `price` DECIMAL(12,4) NULL DEFAULT NULL,
+  `content` TEXT NULL DEFAULT NULL,
+  `created_at` TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP,
+  `edited_at` TIMESTAMP NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
```
Стараемся придерживаться единого наименования полей. updated_at

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+  `content` TEXT NULL DEFAULT NULL,
+  `created_at` TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP,
+  `edited_at` TIMESTAMP NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
+  `rating` FLOAT(10) UNSIGNED NULL DEFAULT NULL,
```
За что отвечает это поле?

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+
```
Нет таблиц:

- избранные пользователи
- уведомления

int64 commented on src/sql/task-force_v2.sql:
```diff
@@ -0,0 +1,185 @@
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`users` (
+  `id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `location_id` INT(10) UNSIGNED NOT NULL,
```
city_id

Commit:
Версия #3, на проверку

AlexBad committed:
Исправления, согласно замечаниям.

Удалены лишние файлы.
Отредактированы Файлы: src/sql/task-force.sql, src/models/Task.php

int64 commented on src/sql/task-force.sql:
```diff
@@ -0,0 +1,216 @@
+CREATE TABLE IF NOT EXISTS `taskforce`.`user_favorites` (
+  `id` INT(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `user_id` INT(10) UNSIGNED NOT NULL,
+  `favorit_id` INT(10) UNSIGNED NOT NULL,
```
тут по имени поля плохо понятно куда оно указывает, лучше сделать что-то вроде favorite_user_id

int64 commented on src/sql/task-force.sql:
```diff
@@ -0,0 +1,216 @@
+ENGINE = InnoDB
+DEFAULT CHARACTER SET = utf8;
+
+CREATE TABLE IF NOT EXISTS `taskforce`.`user_notify` (
```
1. Notify - уведомлять, а нам надо "уведомления" =)
2. Тут должны хранится настройки уведомлений, в пункте 4.9 про это https://up.htmlacademy.ru/yii/1/project/task-force#taskforce-4-9, подумай еще над структурой этой таблицы.

Commit:
Исправления согласно замечаниям

AlexBad committed:
Исправления синтаксических ошибок.

int64 commented on src/sql/task-force.sql:
```diff
@@ -0,0 +1,227 @@
+  `user_id` INT(10) UNSIGNED NOT NULL,
+  `new_message` TINYINT(1) UNSIGNED NOT NULL DEFAULT 1,
+  `new_respond` TINYINT(1) UNSIGNED NOT NULL DEFAULT 1,
+  `task_actions` TINYINT(1) UNSIGNED NOT NULL DEFAULT 1,
```
А почему для уведомлений по заданиям поле одно?

AlexBad replied:
В макете страницы Профиля, указаны только 3 пункта.

AlexBad committed:
Доработка полей таблиц `users` и `task_resonses`

AlexBad committed:
Поле статуса вернул к ENUM типу. Добалвено `Описание` к Заданиям.
